prereise/gather/translate_zones.py

```diff
@@ -0,0 +1,229 @@
+
+    prev_zones = format_zone_df(prev_zones, name_prev)
+    new_zones = format_zone_df(new_zones, name_new)
+
+    # get overlapping areas as new polygons
+    # this drops areas of new zones that do not overlap prev zones
+    # dropped areas will be re-added at the end this function
+    prev_overlapped_by_new = prev_zones.overlay(
+        new_zones, how="identity", keep_geom_type=False
+    )
+
+    # calculate area of overlap as a percentage of the original shape
+    prev_overlapped_by_new["overlap_area"] = prev_overlapped_by_new["geometry"].area
+    prev_overlapped_by_new["pct_of_prev_area"] = (
+        prev_overlapped_by_new["overlap_area"]
+        / prev_overlapped_by_new[f"{name_prev}_area"]
+    )
+
+    # pivots dataframe into a matrix where rows=prev_zones and cols=new_zones
+    prev_to_new_matrix = prev_overlapped_by_new.pivot(
+        index=name_prev, columns=name_new, values="pct_of_prev_area"
+    )
+
+    # clean NaNs
+    prev_to_new_matrix = prev_to_new_matrix.rename(columns={np.nan: "none"})
+    prev_to_new_matrix = prev_to_new_matrix.fillna(value=0)
+
+    # check for prev_zones that do not overlap new_zones, then remove
+    isolated_zones = prev_to_new_matrix.loc[prev_to_new_matrix["none"] >= 1]
+    if len(isolated_zones > 0):
+        print(
+            f"\nWARNING: Detected {name_prev} zones that do not overlap with {name_new}: {list(isolated_zones.index)}"
+        )
+        prev_to_new_matrix = prev_to_new_matrix.loc[prev_to_new_matrix["none"] < 1]
+
+    # Alert if sum > 1 (e.g. new zones overlap each other)
+    # NOTE: only checks to three decimal places
+    prev_to_new_matrix["sum"] = prev_to_new_matrix.sum(axis=1)
+    sum_greater_than_one = prev_to_new_matrix.loc[
+        prev_to_new_matrix["sum"].round(3) > 1
+    ]
+    if len(sum_greater_than_one > 0):
+        print(
+            f"WARNING: Detected rows with a sum greater than one. This most likely means that several {name_new} zones overlap each other."
+        )
+        if verbose:
+            print("\nROWS WITH SUM GREATER THAN ONE:")
+        if verbose:
+            print(sum_greater_than_one.round(3))
+
+    # Scale matrix such that all rows add up to 1 by distributing demand mismatch
+    # proportionally between new zones.
+    # if new zones overlap each other, we may end up with > 100% coverage of old zones
+    # if old zones aren't completely covered by new zones, we have < 100% coverage
+    # Note: we could technically simplify this to just scale_factor = 1/(sum - none) but
+    # the extra cols are nice for verbose mode
+    prev_to_new_matrix["sum_without_none"] = (
+        prev_to_new_matrix["sum"] - prev_to_new_matrix["none"]
+    )
+    prev_to_new_matrix["to_distribute"] = 1 - prev_to_new_matrix["sum_without_none"]
+    prev_to_new_matrix["scale_factor"] = (
+        1 + prev_to_new_matrix["to_distribute"] / prev_to_new_matrix["sum_without_none"]
+    )
```

Comment:
Thanks for putting down detailed explanation in doc strings! Just copying our discussions from Slack to here regarding how to redistribute `over counts` to make rows summing up to 1:

> Another approach to bring it down to 1 is looking at the overlaps among columns and distribute the overlap portion uniformly cross columns that causing duplications rather than all areas.

This will make things messy for sure but gives a relative better projection when dealing with self overlaps. I'm good with the current implementation. Just posting the idea here for future references.

Reply:
I think you're right about this, but it will be quite a bit more work if we want to check for three or more BAs overlapping. To handle three or more simultaneous overlapping BAs we need to recursively check for overlaps until there are none left.

We also need to think about how we handle gaps. For now I think distributing demand proportionally works best to fill in any gaps in the new shapefiles.

Here's an article I've been reading about gaps and overlaps:
https://medium.com/@achm.firmansyah/an-approach-for-checking-overlaps-and-gaps-in-polygons-using-geopandas-ebd6606e7f70